test/unit/gateways/secure_co_test.rb

```diff
@@ -0,0 +1,256 @@
+    assert_success response
+  end
+
+  def test_failed_verify
+    @gateway.expects(:ssl_post).returns(failed_authorize_response)
+    response = @gateway.verify(@credit_card, @options)
+    assert_failure response
+  end
+
+  def test_scrub
+    assert @gateway.supports_scrubbing?
+    assert_equal @gateway.scrub(pre_scrubbed), post_scrubbed
+  end
+
+  def test_optional_fields
+    stub_comms do
+      @gateway.purchase(2000, @credit_card, @options.merge(:currency => 'CAD'))
+    end.check_request do |endpoint, data, headers|
+      assert_match '<requested-amount currency="CAD">20.00</requested-amount>', data
+      assert_match '<account-number>4111111111111111</account-number>', data
+      assert_match '<card-security-code>123</card-security-code>', data
+      assert_match %r{<request-id>[0-9a-f]{32}<\/request-id>}, data
+    end.respond_with(successful_purchase_response)
+
+    stub_comms do
+      @gateway.purchase(2000, @credit_card, @options.merge(
+        ip: '127.0.0.1',
+        request_id: '00000000-0000-0000-0000-000000000000',
+        order_id: 'ORDER1234',
+        custom_fields: [["field1", "value1"], [:field2, "value2"], ["field three", 3]],
+        email: 'test.email@secureco.co',
+      ))
+    end.check_request do |endpoint, data, headers|
+      assert_match '<ip-address>127.0.0.1</ip-address>', data
+      assert_match '<request-id>00000000-0000-0000-0000-000000000000</request-id>', data
+      assert_match '<order-number>ORDER1234</order-number>', data
+      assert_match '<custom-field field-name="field1" field-value="value1"/>', data
+      assert_match '<custom-field field-name="field2" field-value="value2"/>', data
+      assert_match '<custom-field field-name="field three" field-value="3"/>', data
+      assert_match '<email>test.email@secureco.co</email>', data
+    end.respond_with(successful_purchase_response)
+
+    stub_comms do
+      @gateway.purchase(2000, @credit_card, @options.merge(
+        custom_fields: {"field1" => "value1", :field2 => "value2", "field three" => 3},
+      ))
+    end.check_request do |endpoint, data, headers|
+      refute_match %r{<ip-address\b}, data
+      refute_match %r{<order-number\b}, data
+      assert_match '<custom-field field-name="field1" field-value="value1"/>', data
+      assert_match '<custom-field field-name="field2" field-value="value2"/>', data
+      assert_match '<custom-field field-name="field three" field-value="3"/>', data
+      refute_match %r{<email\b}, data
+    end.respond_with(successful_purchase_response)
+  end
+
+  def test_dependent_order_types
+    @gateway.expects(:ssl_post).twice.returns(successful_purchase_response)
+    purchase_response = @gateway.purchase(@amount, @credit_card, @options)
+    stub_comms do
+      @gateway.refund(@amount, purchase_response.authorization, @options)
+    end.check_request do |endpoint, data, headers|
+      assert_match '<transaction-type>refund-purchase</transaction-type>', data
```

Comment:
I am not sure why this is passing on TravisCI but locally I am getting an exception:

```
Failure:
  </<transaction\-type>refund\-purchase<\/transaction\-type>/> expected to be =~
  <"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<payment xmlns=\"http://www.elastic-payments.com/schema/payment\">\n<merchant-account-id>00000000-0000-0000-0000-000000000000</merchant-account-id>\n<transaction-type>authorization</transaction-type>\n<card>\n<account-number>4111111111111111</account-number>\n<card-security-code>123</card-security-code>\n<card-type>visa</card-type>\n<expiration-month>09</expiration-month>\n<expiration-year>2018</expiration-year>\n</card>\n<request-id>6126586b347723ed406120253c43699b</request-id>\n<requested-amount currency=\"AUD\">10.00</requested-amount>\n<entry-mode>ecommerce</entry-mode>\n<account-holder>\n<first-name>Longbob</first-name>\n<last-name>Longsen</last-name>\n</account-holder>\n</payment>\n">.
test_dependent_order_types(SecureCoTest)
cyclotron3k-activemerchant/test/unit/gateways/secure_co_test.rb:169:in `block in test_dependent_order_types'
     166:     stub_comms do
     167:       @gateway.refund(@amount, purchase_response.authorization, @options)
     168:     end.check_request do |endpoint, data, headers|
  => 169:       assert_match '<transaction-type>refund-purchase</transaction-type>', data
     170:     end.respond_with(successful_refund_response)
     171:
     172:     @gateway.expects(:ssl_post).times(3).returns(successful_authorize_response).then.returns(successful_capture_response)
cyclotron3k-activemerchant/test/comm_stub.rb:22:in `block in respond_with'
cyclotron3k-activemerchant/lib/active_merchant/billing/gateways/secure_co.rb:665:in `commit'
cyclotron3k-activemerchant/lib/active_merchant/billing/gateways/secure_co.rb:239:in `authorize'
cyclotron3k-activemerchant/test/unit/gateways/secure_co_test.rb:173:in `test_dependent_order_types'
```